isobus/include/isobus/isobus/can_parameter_group_builder.hpp

```diff
@@ -0,0 +1,503 @@
+#pragma once
+
+#include <vector>
+
+namespace isobus
+{
+	// Note that this class currently only works for packets eight bytes or fewer, because that is
+	// all the use-cases I needed.  The spec has a whole section on how data can span byte
+	// boundaries, and how to align the bits when a data-type that is not a multiple of eight bits
+	// crosses said boundaries.  Then every single parameter group uses padding to avoid those
+	// cases!  The five-bit tractor command types have three bits of padding to make them exactly
+	// one byte.  Why even specify the scheme if it isn't used?
+	//
+	// I wrote a huge amount of this file once, then somehow it vanished.  I don't know how!
+	class ParameterGroupBuilder
+	{
+	private:
+		std::size_t writeOffset = 0;
+		std::size_t readOffset = 0;
+		std::vector<std::uint8_t> buffer;
+
+		std::size_t get_write_byte_offset() const
+		{
+			// Which byte to write to.
+			return writeOffset / 8;
+		}
+
+		std::size_t get_write_bit_offset() const
+		{
+			// Which bit to write to in the current byte.
+			return writeOffset % 8;
```

Comment:
For non-templated functions, prefer putting the implementation in a .cpp file. This will help a number of things such as allowing our static analysis to better check the file, as it will not really be checked right now since the single header is included nowhere.